File: src/main/java/ca/uwindsor/analyzing/CustomStemFilter.java
package ca.uwindsor.analyzing;

import java.io.IOException;
import java.util.Collection;

import org.apache.lucene.analysis.TokenStream;
import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;

/**
 * Filter with stemming for custom terms.
 */
public class CustomStemFilter extends SetTokenFilter
{
    /**
     * Required value for the tokenizing.
     */
    private final CharTermAttribute termAttr = addAttribute(CharTermAttribute.class);

    /**
     * Do not add any terms and creates a hash set to store term.
     *
     * @param input The input stream.
     */
    public CustomStemFilter(TokenStream input)
    {
        super(input);
    }

    /**
     * Start with an existing set.
     *
     * @param input The input stream.
     * @param set   The starting set.
     */
    public CustomStemFilter(TokenStream input, Collection<String> set)
    {
        super(input, set);
    }

    /**
     * Add terms from a file which will be normalized into a hash set.
     *
     * @param input The input stream.
     * @param path  The path to the file.
     */
    public CustomStemFilter(TokenStream input, String path)
    {
        super(input, path);
    }

    /**
     * Add terms from a file into a hash set.
     *
     * @param input     The input stream.
     * @param path      The path to the file.
     * @param normalize If the terms should be normalized or not.
     */
    public CustomStemFilter(TokenStream input, String path, Boolean normalize)
    {
        super(input, path, normalize);
    }

    /**
     * Add terms from a file which will be normalized into an existing set.
     *
     * @param input The input stream.
     * @param set   The starting set.
     * @param path  The path to the file.
     */
    public CustomStemFilter(TokenStream input, Collection<String> set, String path)
    {
        super(input, set, path);
    }

    /**
     * Add terms from a file into an existing set.
     *
     * @param input     The input stream.
     * @param set       The starting set.
     * @param path      The path to the file.
     * @param normalize If the terms should be normalized or not.
     */
    public CustomStemFilter(TokenStream input, Collection<String> set, String path, Boolean normalize)
    {
        super(input, set, path, normalize);
    }

    /**
     * See if we should increment the number of tokens.
     *
     * @return True if a stemmed term was matched, false otherwise.
     * @throws IOException Reading the input fails.
     */
    @Override
    public final boolean incrementToken() throws IOException
    {
        // First, check if there is another token to process.
        if (!input.incrementToken())
        {
            return false;
        }

        String term = termAttr.toString();

        // Check if the term is in the custom stems set
        if (term == null || term.isEmpty())
        {
            return false;
        }

        // Try to step the term.
        String stem = StringStartsWith(term, false);

        // If there was no stem, just keep the term the same.
        if (stem == null)
        {
            stem = term;
        }

        termAttr.setEmpty().append(stem);

        return true;
    }
}

File: Lucene/Analyzing/CustomStemFilter.cs
using Lucene.Net.Analysis;
using Lucene.Net.Analysis.TokenAttributes;
using SearchEngine.Lucene.Utility;

namespace SearchEngine.Lucene.Analyzing;

/// <summary>
/// Custom stemming filter.
/// </summary>
public sealed class CustomStemFilter : TokenFilter
{
    /// <summary>
    /// Required to help with building terms.
    /// </summary>
    private readonly ICharTermAttribute _termAttr;
    
    /// <summary>
    /// The stems.
    /// </summary>
    private readonly TermsCollection _stems;
    
    /// <summary>
    /// Create the filter.
    /// </summary>
    /// <param name="input">The token stream.</param>
    /// <param name="collection">The stems.</param>
    internal CustomStemFilter(TokenStream input, ICollection<string> collection) : base(input)
    {
        _stems = new(collection);
        _termAttr = AddAttribute<ICharTermAttribute>();
    }

    /// <summary>
    /// Look for the next token.
    /// </summary>
    /// <returns>True if we still have terms, false otherwise</returns>
    public override bool IncrementToken()
    {
        // If there are no more tokens, there is nothing left to do.
        if (!m_input.IncrementToken())
        {
            return false;
        }

        // If our existing string is empty, there is nothing to do.
        string term = _termAttr.ToString();
        if (term.Length < 1)
        {
            return false;
        }

        // See if we can get a customs stem, or otherwise keep the term the same.
        _termAttr.SetEmpty().Append(_stems.StringStartsWith(term, false) ?? term);
        return true;
    }
}